### PR TITLE
 Add list-based composite auto-configuration

### PR DESCRIPTION
**Pull request by pivotal-dylan-roberts**
Add support for listing environment repository configuration blocks under spring.cloud.config.server.composite, to create a composite environment repository.

Fixes gh-616

### CODE REVIEW

**Comment by pivotal-issuemaster** [comment]
@pivotal-dylan-roberts Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-config&pullRequestId=912)!

[Click here](https://cla.pivotal.io/sync/spring?repositoryId=spring-cloud/spring-cloud-config&pullRequestId=912) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

**Comment by pivotal-issuemaster** [comment]
@pivotal-dylan-roberts Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/spring?repositoryId=spring-cloud/spring-cloud-config&pullRequestId=912)!

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java [hunk 1]
```diff
@@ -18,38 +18,36 @@
...
-import org.springframework.cloud.config.server.environment.MultipleJGitEnvironmentRepository;
-import org.springframework.cloud.config.server.environment.NativeEnvironmentRepository;
-import org.springframework.cloud.config.server.environment.SvnKitEnvironmentRepository;
-import org.springframework.cloud.config.server.environment.VaultEnvironmentRepository;
-import org.springframework.context.annotation.Bean;
-import org.springframework.context.annotation.Configuration;
-import org.springframework.context.annotation.Import;
-import org.springframework.context.annotation.Profile;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.cloud.config.server.environment.*;
```
Comment (spencergibb) [inline spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java]: Please, no * imports.

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java [hunk 2]
```diff
@@ -18,38 +18,36 @@
...
+import org.springframework.context.annotation.*;
 import org.springframework.core.env.ConfigurableEnvironment;
+import org.springframework.core.env.Environment;
 import org.springframework.jdbc.core.JdbcTemplate;
 import org.springframework.web.client.RestTemplate;
 
+import java.util.*;
```
Comment (spencergibb) [inline spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java]: Please use our import order

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java [hunk 3]
```diff
@@ -80,8 +78,8 @@ public EnvironmentWatch environmentWatch() {
 
 @Configuration
 @ConditionalOnMissingBean(EnvironmentRepository.class)
+@Profile("!composite")
```
Comment (spencergibb) [inline spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java]: You'll have to explain this. Seems counter-intuitive to me.

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java [hunk 4]
```diff
@@ -139,32 +138,88 @@ public NativeEnvironmentRepository nativeEnvironmentRepository() {
...
+@Profile("composite")
+class CompositeRepositoryConfiguration {
+
+	@Bean
+	public Map<String, EnvironmentRepositoryFactory<?,?>> environmentRepositoryFactories(
+			ConfigurableEnvironment environment, ConfigServerProperties server,
+			Optional<TransportConfigCallback> transportConfigCallback, HttpServletRequest request,
+			EnvironmentWatch watch, JdbcTemplate jdbc) {
+		Map<String, EnvironmentRepositoryFactory<?,?>> factories = new HashMap<>();
+		factories.put("git", new GitEnvironmentRepositoryFactory(environment, server, transportConfigCallback));
```
Comment (spencergibb) [inline spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java]: I don't think a static list of factories is flexible enough. How would custom repositories work?

Comment (spencergibb) [reply spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java]: The `EnvironmentRepositoryFactory`s should be beans that get injected. Spring has a [`FactoryBean<T>`] (https://docs.spring.io/spring/docs/current/javadoc-api/org/springframework/beans/factory/config/AbstractFactoryBean.html) that should probably be used. @dsyer would know more about that.

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/AbstractScmEnvironmentProperties.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package org.springframework.cloud.config.server.environment;
+
+import org.springframework.cloud.config.server.support.AbstractScmAccessorProperties;
+
+public abstract class AbstractScmEnvironmentProperties extends AbstractScmAccessorProperties {
```
Comment (spencergibb) [inline spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/AbstractScmEnvironmentProperties.java]: I'd say push this down to AbstractScmAccessorProperties

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/JdbcEnvironmentRepository.java [hunk 6]
```diff
@@ -61,7 +59,15 @@
 	public JdbcEnvironmentRepository(JdbcTemplate jdbc) {
 		this.jdbc = jdbc;
 	}
-	
+
+	public JdbcEnvironmentRepository(JdbcTemplate jdbc, JdbcEnvironmentProperties properties) {
+		this.jdbc = jdbc;
+		if (properties != null) {
```
Comment (spencergibb) [inline spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/JdbcEnvironmentRepository.java]: I'd say in general assume not null

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/JdbcEnvironmentRepository.java [hunk 7]
```diff
@@ -61,7 +59,15 @@
...
+			this.order = properties.getOrder() == null ? this.order : properties.getOrder();
```
Comment (spencergibb) [inline spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/JdbcEnvironmentRepository.java]: Just set it, no need for null check.

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/SvnKitEnvironmentProperties.java [hunk 8]
```diff
@@ -0,0 +1,16 @@
+package org.springframework.cloud.config.server.environment;
+
+import org.springframework.boot.context.properties.ConfigurationProperties;
+
+@ConfigurationProperties("spring.cloud.config.server.svn")
+public class SvnKitEnvironmentProperties extends AbstractScmEnvironmentProperties {
+    private String defaultLabel;
+
+    public String getDefaultLabel() {
```
Comment (spencergibb) [inline spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/SvnKitEnvironmentProperties.java]: Looks like the DefaultLabel can be pushed down.

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/JdbcEnvironmentRepository.java (hunk 7 above)
Comment (pivotal-dylan-roberts) [reply spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/JdbcEnvironmentRepository.java]: Would you prefer to see the defaults for all fields on the environment repositories set in the properties classes?

Comment (spencergibb) [reply spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/JdbcEnvironmentRepository.java]: All, I forgot there are defaults, but those should now go in the properties class

Comment (pivotal-dylan-roberts) [reply spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/environment/JdbcEnvironmentRepository.java]: @spencergibb Cool, will do!

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java (hunk 3 above)
Comment (pivotal-dylan-roberts) [reply spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java]: I'm finding that evaluation of `@ConditionalOnMissingBean` is happening before the `EnvironmentRepository`s exist.

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java (hunk 4 above)
Comment (pivotal-dylan-roberts) [reply spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java]: @spencergibb I've added dynamic factory resolution and also added a test to show how custom repositories would work.

**Comment by spencergibb** [comment]
Awesome. I'll review as soon as I can, traveling

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java (hunk 3 above)
Comment (spencergibb) [reply spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java]: the order of the `@Import` matters. You've put composite after default.

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/composite/GitCompositeEnvironmentRepositoryFactory.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+public class GitCompositeEnvironmentRepositoryFactory extends
+		AbstractCompositeEnvironmentRepositoryFactory<MultipleJGitEnvironmentRepository, MultipleJGitEnvironmentProperties> {
+
+	public GitCompositeEnvironmentRepositoryFactory(EnvironmentRepositoryFactory<MultipleJGitEnvironmentRepository,
+			MultipleJGitEnvironmentProperties> delegate, Environment environment) {
+		super(delegate, environment);
+	}
+
+	@Override
+	public MultipleJGitEnvironmentRepository build(int index) throws Exception {
```
Comment (spencergibb) [inline spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/composite/GitCompositeEnvironmentRepositoryFactory.java]: needed?

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/composite/JdbcCompositeEnvironmentRepositoryFactory.java [hunk 10]
```diff
@@ -0,0 +1,25 @@
...
+public class JdbcCompositeEnvironmentRepositoryFactory extends
+		AbstractCompositeEnvironmentRepositoryFactory<JdbcEnvironmentRepository, JdbcEnvironmentProperties> {
+
+	public JdbcCompositeEnvironmentRepositoryFactory(EnvironmentRepositoryFactory<JdbcEnvironmentRepository,
+			JdbcEnvironmentProperties> delegate, Environment environment) {
+		super(delegate, environment);
+	}
+
+	@Override
+	public JdbcEnvironmentRepository build(int index) throws Exception {
```
Comment (spencergibb) [inline spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/composite/JdbcCompositeEnvironmentRepositoryFactory.java]: Needed?

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/composite/NativeCompositeEnvironmentRepositoryFactory.java [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+public class NativeCompositeEnvironmentRepositoryFactory extends
+		AbstractCompositeEnvironmentRepositoryFactory<NativeEnvironmentRepository, NativeEnvironmentProperties> {
+
+	public NativeCompositeEnvironmentRepositoryFactory(EnvironmentRepositoryFactory<NativeEnvironmentRepository,
+			NativeEnvironmentProperties> delegate, Environment environment) {
+		super(delegate, environment);
+	}
+
+	@Override
+	public NativeEnvironmentRepository build(int index) throws Exception {
```
Comment (spencergibb) [inline spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/composite/NativeCompositeEnvironmentRepositoryFactory.java]: Needed?

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/composite/SvnCompositeEnvironmentRepositoryFactory.java [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+public class SvnCompositeEnvironmentRepositoryFactory extends
+		AbstractCompositeEnvironmentRepositoryFactory<SvnKitEnvironmentRepository, SvnKitEnvironmentProperties> {
+
+	public SvnCompositeEnvironmentRepositoryFactory(EnvironmentRepositoryFactory<SvnKitEnvironmentRepository,
+			SvnKitEnvironmentProperties> delegate, Environment environment) {
+		super(delegate, environment);
+	}
+
+	@Override
+	public SvnKitEnvironmentRepository build(int index) throws Exception {
```
Comment (spencergibb) [inline spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/composite/SvnCompositeEnvironmentRepositoryFactory.java]: Needed?

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java [hunk 13]
```diff
@@ -92,32 +128,31 @@ public EnvironmentWatch environmentWatch() {
...
-	public MultipleJGitEnvironmentRepository defaultEnvironmentRepository() {
-		MultipleJGitEnvironmentRepository repository = new MultipleJGitEnvironmentRepository(
-				this.environment);
-		repository.setTransportConfigCallback(this.transportConfigCallback);
-		if (this.server.getDefaultLabel() != null) {
-			repository.setDefaultLabel(this.server.getDefaultLabel());
-		}
-		return repository;
+	public MultipleJGitEnvironmentRepository defaultEnvironmentRepository(
+			Optional<MultipleJGitEnvironmentProperties> environmentProperties) throws Exception {
```
Comment (spencergibb) [inline spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java]: The properties shouldn't be `Optional` should they?

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java [hunk 14]
```diff
@@ -155,16 +187,90 @@ public SvnKitEnvironmentRepository svnKitEnvironmentRepository() {
...
+
+	@Bean
+	public boolean compositeTypesIncludeSearchPathLocator(
+			Environment environment, ConfigurableListableBeanFactory beanFactory) throws ClassNotFoundException {
+		return CompositeUtils.compositeTypesIncludeSearchPathLocator(environment, beanFactory);
+	}
+
+	@Primary
+	@Bean
+	@ConditionalOnExpression("#{ compositeTypesIncludeSearchPathLocator }")
```
Comment (spencergibb) [inline spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java]: This is very unusual

Comment (spencergibb) [reply spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java]: I'd do a custom `@Conditional`. That way you don't need a boolean bean.

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java [hunk 15]
```diff
@@ -92,32 +128,31 @@ public EnvironmentWatch environmentWatch() {
...
-				this.environment);
-		repository.setTransportConfigCallback(this.transportConfigCallback);
-		if (this.server.getDefaultLabel() != null) {
-			repository.setDefaultLabel(this.server.getDefaultLabel());
-		}
-		return repository;
+	public MultipleJGitEnvironmentRepository defaultEnvironmentRepository(
+			Optional<MultipleJGitEnvironmentProperties> environmentProperties) throws Exception {
+		MultipleJGitEnvironmentRepositoryFactory gitEnvironmentRepositoryFactory =
+				new MultipleJGitEnvironmentRepositoryFactory(environment, server,
```
Comment (spencergibb) [inline spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java]: Can you explain why there are two different factories for jgit?

File: spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java (hunk 3 above)
Comment (pivotal-dylan-roberts) [reply spring-cloud-config-server/src/main/java/org/springframework/cloud/config/server/config/EnvironmentRepositoryConfiguration.java]: Thanks @spencergibb, that worked.

**Comment by pivotal-dylan-roberts** [comment]
@spencergibb Managed to sort out all the issues you mentioned. Please review at your convenience!

**Pull request merged**